### PR TITLE
Staging

### PR DESCRIPTION
**Pull request by simon-norman**

### CODE REVIEW

File: src/main/kotlin/com/api_forty_winks/security/WebSecurityConfig.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.http.HttpMethod
+import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
+import org.springframework.security.config.annotation.web.builders.HttpSecurity
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
+import org.springframework.security.core.userdetails.UserDetailsService
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
+import com.auth0.spring.security.api.JwtWebSecurityConfigurer
+import org.springframework.beans.factory.annotation.Value
+
```
Comment (simon-norman) [inline src/main/kotlin/com/api_forty_winks/security/WebSecurityConfig.kt:14]: Any imports unused?

File: src/main/kotlin/com/api_forty_winks/shelter/ShelterList.kt [hunk 2]
```diff
@@ -5,18 +5,89 @@ class ShelterList {
...
+        val shelterData0 = mutableMapOf<String,String>()
+        shelterData0["name"] = "Salvation Army - Cambria House"
+        shelterData0["price"] = "3.00 per night"
+        shelterData0["description"] = "Single homeless men aged 18+ with low to medium support needs who require assistance with developing independent living skills. 2 beds for single homeless men who are registered disabled."
+        shelterData0["location"] = "37 Hunter Street, London, WC1N 1BJ"
+        shelterData0["mapurl"] = "https://www.google.com/maps/embed?pb=!1m18!1m12!1m3!1d2482.3113043813714!2d-0.12599658473005426!3d51.52584981717579!2m3!1f0!2f0!3f0!3m2!1i1024!2i768!4f13.1!3m3!1m2!1s0x48761b3a046ac8c7%3A0x67e9011515119add!2sHunter+St%2C+Bloomsbury%2C+London+WC1N+1BJ!5e0!3m2!1sen!2suk!4v1555761182587!5m2!1sen!2suk"
+        sheltersRaw.add(shelterData0)
+
+        println(sheltersRaw)
+
```
Comment (simon-norman) [inline src/main/kotlin/com/api_forty_winks/shelter/ShelterList.kt:90]: Can remove println

File: src/main/kotlin/com/api_forty_winks/voucher/Voucher.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+@Entity
+@Table(name = "voucher")
+@Embeddable
+class Voucher(
+    @Id @GeneratedValue(strategy = GenerationType.IDENTITY) @Column(columnDefinition = "serial") var code: Long,
+    var amount: Float? = null
+) {
+    fun Voucher() {}
+}
```
Comment (simon-norman) [inline src/main/kotlin/com/api_forty_winks/voucher/Voucher.kt:13]: Can this default constructor fun Voucher() be removed now? Test

File: src/main/kotlin/com/api_forty_winks/voucher/VoucherController.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    fun newVoucher(@RequestBody voucher: Voucher) : Voucher {
+        voucherRepository.save(voucher)
+        return voucher
+    }
+
+    @PostMapping("private/vouchers/1/redemption")
+    @CrossOrigin
+    fun redeemVoucher() : ResponseEntity<Unit> {
+        return ResponseEntity.ok().build()
+    }
```
Comment (simon-norman) [inline src/main/kotlin/com/api_forty_winks/voucher/VoucherController.kt:20]: Redemption to be added

File: src/main/kotlin/com/api_forty_winks/voucher/VoucherController.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+    @GetMapping("/voucher")
+    @CrossOrigin
+    fun getVoucher(@RequestParam("code") code : Long) : Voucher {
+        var v = voucherRepository.getOne(code)
+        return v
+    }
```
Comment (simon-norman) [inline src/main/kotlin/com/api_forty_winks/voucher/VoucherController.kt:27]: Can rename v to voucher?

File: src/main/resources/application.properties [hunk 6]
```diff
@@ -0,0 +1,11 @@
+## Spring DATASOURCE (DataSourceAutoConfiguration & DataSourceProperties)
+spring.datasource.url =jdbc:postgresql://localhost:5432/forty_winks_dev
+spring.datasource.username =student
+spring.datasource.password =''
+# Disable feature detection by this undocumented parameter. Check the org.hibernate.engine.jdbc.internal.JdbcServiceImpl.configure method for more details.
```
Comment (simon-norman) [inline src/main/resources/application.properties:5]: Probably should move auth0 audience and issuer into environment variables

File: src/test/kotlin/com/api_forty_winks/message/MessageControllerApiTests.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+@RunWith(SpringRunner::class)
+@SpringBootTest
+@AutoConfigureMockMvc
+
+class MessageControllerApiTests {
+    @Test
+    fun `It successfully sends a text message`(){
+        val requestParams = JSONObject()
+        requestParams["number"] = "+447722066721"
+        requestParams["code"] = "FW5"
```
Comment (simon-norman) [inline src/test/kotlin/com/api_forty_winks/message/MessageControllerApiTests.kt:23]: Can change this number (is James') to fake one.

File: src/test/kotlin/com/api_forty_winks/message/MessageTests.kt [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+import org.mockito.BDDMockito.given
+import org.mockito.Mockito
+import org.mockito.Mockito.verify
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.test.context.junit4.SpringRunner
+
+
+@RunWith(SpringRunner::class)
+@SpringBootTest
+
```
Comment (simon-norman) [inline src/test/kotlin/com/api_forty_winks/message/MessageTests.kt:17]: If time, flesh out to full test of message

File: src/main/resources/application.properties (hunk 6 above)
Comment (limavadakara) [reply src/main/resources/application.properties:5]: Addition of environment variables to store Db details messes up the tests as it says that an application context is missing. So this is a bit more involved. So will look at it later.

Comment (simon-norman) [reply src/main/resources/application.properties:5]: Ya is possible to do but actually these variables are not sensitive as just refer to local db instances - heroku has its own so not live or staging

**Pull request merged**